File: 1406-stone-game-iii/1406-stone-game-iii.c
//The first code is solved by using recursion method


#define max(a, b) ((a) > (b) ? (a) : (b))
// int solve(int* stoneValue , int i , int n , int* t){
//     if(i >= n){
//         return 0;
//     }

//     if(t[i] != -1){
//         return t[i];
//     }

//     int result = stoneValue[i] - solve(stoneValue , i + 1 , n , t);

//     if(i + 1 < n){
//         result = max(result , stoneValue[i] + stoneValue[i + 1] - solve(stoneValue , i + 2 , n ,t));
//     }

//     if(i + 2 < n){
//         result = max(result , stoneValue[i] + stoneValue[i + 1] + stoneValue[i + 2] - solve(stoneValue , i + 3 , n , t));
//     }

//     return t[i] = result;
// }

// char* stoneGameIII(int* stoneValue, int stoneValueSize) {
//     int memo[stoneValueSize + 1];
//     memset(memo , -1 , sizeof(memo));
//     int diff = solve(stoneValue , 0 , stoneValueSize, memo);

//     if(diff > 0){
//         return "Alice";
//     }else if(diff < 0){
//         return "Bob";
//     }
//     return "Tie";
// }
char* stoneGameIII(int* stoneValue, int stoneValueSize){
    int memo[stoneValueSize + 1];
    memset(memo , -1 , sizeof(int));
    int a = 0;//i + 1
    int b = 0;//i + 2
    int c = 0;//i + 3
    for(int i = stoneValueSize - 1; i >= 0; i--){
        int result = INT_MIN;
        result = max(result , stoneValue[i] - a);
        if(i + 1 < stoneValueSize){
            result = max(result , stoneValue[i] + stoneValue[i + 1] - b);
        }
        if(i + 2 < stoneValueSize){
            result = max(result , stoneValue[i] + stoneValue[i + 1] + stoneValue[i + 2] - c);
        }
        c = b; 
        b = a;
        a = result;
    }
    int diff = a;
    if(diff > 0){
        return "Alice";
    }
    if(diff < 0){
        return "Bob";
    }
    return "Tie";
}

File: 1406-stone-game-iii/1406-stone-game-iii.py
# My first approach was using recursion method

# class Solution:
#     def stoneGameIII(self, stoneValue: List[int]) -> str:
#         n = len(stoneValue)
#         @lru_cache(None)
#         def solve(i: int) -> int:
#             if i >= n:
#                 return 0
#             res = stoneValue[i] - solve(i + 1)
#             if i + 1 < n:
#                 res = max(res , (stoneValue[i] + stoneValue[i + 1] - solve(i + 2)))
#             if i + 2 < n:
#                 res = max(res , (stoneValue[i] + stoneValue[i + 1] + stoneValue[ i + 2] - solve(i + 3)))
#             return res
#         diff = solve(0)
#         if diff > 0:
#             return "Alice"
#         elif diff < 0:
#             return "Bob"
#         return "Tie"
class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        a = 0
        b = 0
        c = 0
        for i in range(n - 1 , -1 , -1):
            result = stoneValue[i] - a
            if i + 1 < n:
                result = max(result , (stoneValue[i] + stoneValue[i + 1] - b))
            if i + 2 < n:
                result = max(result , (stoneValue[i] + stoneValue[i + 1] + stoneValue[i + 2] - c))
            c = b
            b = a
            a = result
        diff = a
        if diff > 0:
            return "Alice"
        elif diff < 0:
            return "Bob"
        return "Tie"